### PR TITLE
add prioritizer with weight

### PR DESCRIPTION
**Pull request by haoqing0110**
Ref: https://github.com/open-cluster-management/backlog/issues/15599

Signed-off-by: haoqing0110 <qhao@redhat.com>

### CODE REVIEW

File: pkg/controllers/scheduling/schedule.go [hunk 1]
```diff
@@ -145,16 +147,24 @@ func (s *pluginScheduler) Schedule(
 			return nil, err
 		}
 
-		results.scoreRecords = append(results.scoreRecords, PriorizeResult{Name: p.Name(), Scores: score})
+		// Get the prioritizer weight. Weight defined in each placement overwrites
```
Comment (qiujian16) [inline pkg/controllers/scheduling/schedule.go:150]: overwrites --> overrides

File: pkg/controllers/scheduling/schedule.go [hunk 2]
```diff
@@ -145,16 +147,24 @@ func (s *pluginScheduler) Schedule(
...
+		// the default value defined in NewPluginScheduler().
+		weight := p.Weight()
+		if customizeWeight, err := getPlacementPrioritizerWeight(placement, p.Name()); err == nil {
```
Comment (qiujian16) [inline pkg/controllers/scheduling/schedule.go:153]: I think you  would want to evaluate the weight in placement at first and put it in a map at first.  So we do not need to iterate here.

File: pkg/helpers/testing/builders.go [hunk 3]
```diff
@@ -36,6 +36,14 @@ func (b *placementBuilder) WithNOC(noc int32) *placementBuilder {
 	return b
 }
 
+func (b *placementBuilder) WithPrioritizerCConfig(name string, weight int32) *placementBuilder {
```
Comment (qiujian16) [inline pkg/helpers/testing/builders.go:39]: typo here?

File: pkg/plugins/balance/balance.go [hunk 4]
```diff
@@ -21,21 +22,29 @@ const (
...
 func (b *Balance) Name() string {
-	return "balance"
+	return reflect.TypeOf(*b).Name()
 }
 
 func (b *Balance) Description() string {
 	return description
 }
 
+func (b *Balance) Weight() int32 {
```
Comment (qiujian16) [inline pkg/plugins/balance/balance.go:44]: based on the discussion with @elgnay I think we do not want plugin to  provide this interface. The weight should purely controlled by  scheduler.

**Review by qiujian16 — COMMENTED** [review]
i think this is a  clear PR. But i would like it to be:

1. define the defaultWeight (map[string]int) in schedule.go with  some comments.
2. override the defaultWeight by evaluating the placement each time.

File: pkg/plugins/balance/balance.go (hunk 4 above)
Comment (haoqing0110) [reply pkg/plugins/balance/balance.go:44]: ok, will remove it.

File: pkg/controllers/scheduling/schedule.go [hunk 5]
```diff
@@ -48,6 +49,7 @@ type FilterResult struct {
 
 type PriorizeResult struct {
```
Comment (elgnay) [inline pkg/controllers/scheduling/schedule.go:50]: Please also help to correct some typos in schedule.go:
- PriorizeResult --> PrioritizerResult
- PrioritizeSore --> PrioritizerScore
- PriorizeResults() --> PrioritizerResults()

File: pkg/helpers/testing/builders.go (hunk 3 above)
Comment (haoqing0110) [reply pkg/helpers/testing/builders.go]: fixed.

File: pkg/controllers/scheduling/schedule.go (hunk 2 above)
Comment (haoqing0110) [reply pkg/controllers/scheduling/schedule.go]: fixed.

File: pkg/controllers/scheduling/schedule.go (hunk 1 above)
Comment (haoqing0110) [reply pkg/controllers/scheduling/schedule.go]: fixed.

File: pkg/controllers/scheduling/schedule.go (hunk 5 above)
Comment (haoqing0110) [reply pkg/controllers/scheduling/schedule.go]: fixed.

File: pkg/controllers/scheduling/schedule.go [hunk 6]
```diff
@@ -204,6 +221,25 @@ func selectClusters(placement *clusterapiv1alpha1.Placement, clusters []*cluster
 	return decisions
 }
 
+// Get the placement's prioritizer weights as a map.
+func getPlacementPrioritizerWeights(placement *clusterapiv1alpha1.Placement) map[string]int32 {
```
Comment (qiujian16) [inline pkg/controllers/scheduling/schedule.go:225]: I would combine this with getWeight, like

```go
weight := make(map[string]int32)
// copy the default weight
for name, weight := range s.prioritizerWeights {
    weight[name] = weight
}

// override the default weight
for _, pc := range placement.Spec.PrioritizerConfigs {
		weight[pc.Name] = pc.Weight
}

```

Comment (haoqing0110) [reply pkg/controllers/scheduling/schedule.go]: Thanks, update the code.

File: pkg/controllers/scheduling/scheduling_controller.go [hunk 7]
```diff
@@ -470,5 +472,15 @@ func (c *schedulingController) createOrUpdatePlacementDecision(
 		"DecisionUpdate", "DecisionUpdated",
 		"Decision %s is updated with placement %s in namespace %s", placementDecision.Name, placement.Name, placement.Namespace)
 
+	// update the event with prioritizer score.
```
Comment (qiujian16) [inline pkg/controllers/scheduling/scheduling_controller.go:475]: I think what you have here should be each cluster's total score in the decision, right?

Comment (haoqing0110) [reply pkg/controllers/scheduling/scheduling_controller.go:475]: yes, change to update cluster total score to event as below. 
```
Events:
  Type    Reason          Age   From                 Message
  ----    ------          ----  ----                 -------
  Normal  DecisionCreate  49m   placementController  Decision placement1-decision-1 is created with placement placement1 in namespace default
  Normal  DecisionUpdate  49m   placementController  Decision placement1-decision-1 is updated with placement placement1 in namespace default
  Normal  ScoreUpdate     49m   placementController  prod-9:100 prod-10:100 prod-6:100 prod-8:100 prod-7:100
```

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**